File: src/med_exam_toolkit/dedup.py
from __future__ import annotations
import hashlib
import re
from med_exam_toolkit.models import Question


def _normalize_text(text: str) -> str:
    """去除空白、标点差异，统一用于指纹计算"""
    text = re.sub(r"\s+", "", text)
    # 统一中英文标点
    text = text.replace("，", ",").replace("。", ".").replace("；", ";")
    text = text.replace("：", ":").replace("（", "(").replace("）", ")")
    return text.lower()

def _resolve_answer_text(sq) -> str:
    """将答案字母转为实际选项文本，避免选项顺序影响指纹"""
    try:
        idx = ord(sq.answer.strip().upper()) - ord("A")
        if 0 <= idx < len(sq.options):
            return _normalize_text(sq.options[idx])
    except (TypeError, ValueError):
        pass
    return sq.answer.strip().upper()

def compute_fingerprint(q: Question, strategy: str = "strict") -> str:
    """
    计算题目指纹。

    strategy:
        - content: 仅基于题干/子题文本
        - strict:  题干 + 选项(排序) + 答案文本 + 共享选项(B型题)
    """
    parts: list[str] = []

    # 共享题干
    if q.stem:
        parts.append(_normalize_text(q.stem))

    # B型题的共享选项（必须参与指纹计算，避免错误去重）
    if q.shared_options and strategy == "strict":
        sorted_shared = sorted(_normalize_text(opt) for opt in q.shared_options)
        parts.extend(sorted_shared)

    for sq in q.sub_questions:
        parts.append(_normalize_text(sq.text))
        if strategy == "strict":
            # 排序选项，消除顺序差异
            sorted_opts = sorted(_normalize_text(opt) for opt in sq.options)
            parts.extend(sorted_opts)
            # 用答案文本而非字母
            parts.append(_resolve_answer_text(sq))

    raw_str = "|".join(parts)
    return hashlib.sha256(raw_str.encode("utf-8")).hexdigest()[:16]


def deduplicate(
    questions: list[Question],
    strategy: str = "strict",
) -> list[Question]:
    """
    去重，返回去重后的列表。
    保留首次出现的题目，后续重复的丢弃。
    """
    seen: dict[str, Question] = {}
    duplicates = 0

    for q in questions:
        fp = compute_fingerprint(q, strategy)
        q.fingerprint = fp
        if fp in seen:
            duplicates += 1
            # 合并来源信息
            existing = seen[fp]
            if q.pkg not in existing.pkg:
                existing.pkg += f",{q.pkg}"
        else:
            seen[fp] = q

    result = list(seen.values())
    print(f"[INFO] 去重完成: {len(questions)} -> {len(result)} (去除 {duplicates} 条重复)")
    return result

File: internal/dedup/dedup.go
package dedup

import (
	"crypto/sha256"
	"fmt"
	"strings"
	"unicode"

	"github.com/med-exam-kit/med-exam-kit/internal/models"
)

// Deduplicate removes duplicate questions, keeping the first occurrence.
// Duplicates have their pkg merged into the surviving question's Pkg field.
// strategy: "strict" (default) or "content".
func Deduplicate(questions []*models.Question, strategy string) []*models.Question {
	if strategy == "" {
		strategy = "strict"
	}
	seen := make(map[string]*models.Question, len(questions))
	order := make([]string, 0, len(questions))
	dupes := 0

	for _, q := range questions {
		fp := ComputeFingerprint(q, strategy)
		q.Fingerprint = fp
		if existing, ok := seen[fp]; ok {
			dupes++
			// Merge source pkg using exact set membership
			existingPkgs := pkgSet(existing.Pkg)
			if _, already := existingPkgs[q.Pkg]; !already && q.Pkg != "" {
				existing.Pkg += "," + q.Pkg
			}
		} else {
			seen[fp] = q
			order = append(order, fp)
		}
	}

	result := make([]*models.Question, 0, len(order))
	for _, fp := range order {
		result = append(result, seen[fp])
	}
	_ = dupes
	return result
}

// ComputeFingerprint calculates a 16-char hex fingerprint for a question.
func ComputeFingerprint(q *models.Question, strategy string) string {
	parts := make([]string, 0, 16)

	if q.Stem != "" {
		parts = append(parts, normalise(q.Stem))
	}
	if len(q.SharedOptions) > 0 && strategy == "strict" {
		sorted := sortedNorm(q.SharedOptions)
		parts = append(parts, sorted...)
	}
	for _, sq := range q.SubQuestions {
		parts = append(parts, normalise(sq.Text))
		if strategy == "strict" {
			parts = append(parts, sortedNorm(sq.Options)...)
			parts = append(parts, resolveAnswerText(&sq))
		}
	}

	raw := strings.Join(parts, "|")
	sum := sha256.Sum256([]byte(raw))
	return fmt.Sprintf("%x", sum[:8]) // 16 hex chars
}

// ── helpers ────────────────────────────────────────────────────────────

func normalise(s string) string {
	// collapse whitespace
	s = strings.Join(strings.Fields(s), "")
	// unify common punctuation variants
	replacer := strings.NewReplacer(
		"，", ",", "。", ".", "；", ";",
		"：", ":", "（", "(", "）", ")",
	)
	s = replacer.Replace(s)
	return strings.Map(unicode.ToLower, s)
}

func sortedNorm(opts []string) []string {
	normed := make([]string, len(opts))
	for i, o := range opts {
		normed[i] = normalise(o)
	}
	// simple insertion sort (lists are short)
	for i := 1; i < len(normed); i++ {
		for j := i; j > 0 && normed[j] < normed[j-1]; j-- {
			normed[j], normed[j-1] = normed[j-1], normed[j]
		}
	}
	return normed
}

func resolveAnswerText(sq *models.SubQuestion) string {
	ans := strings.TrimSpace(sq.Answer)
	if len(ans) == 1 {
		idx := int(ans[0]|0x20) - int('a') // case-insensitive
		if idx >= 0 && idx < len(sq.Options) {
			return normalise(sq.Options[idx])
		}
	}
	return strings.ToUpper(ans)
}

func pkgSet(pkg string) map[string]struct{} {
	parts := strings.Split(pkg, ",")
	m := make(map[string]struct{}, len(parts))
	for _, p := range parts {
		m[strings.TrimSpace(p)] = struct{}{}
	}
	return m
}
